torch/distributed/_sharded_tensor/api.py

```diff
@@ -751,3 +928,122 @@ def _create_tensor_from_params(*size, local_device, tensor_init_params: TensorIn
                           device=local_device, )
     else:
         raise ValueError(f'Unsupported create_op: {tensor_init_params.create_op}')
+
+
+class _PartialTensor(object):
+    """
+    PartialTensor is an abstraction to represent Tensors that need
+    aggregation across multiple devices and multiple processes.
+
+    PartialTensor is initialized in an SPMD like fashion where each rank
+    initializes the PartialTensor. The PartialTensor object on each rank
+    then only stores the local partial shard, process group and the
+    aggregation way to get a full tensor.
+
+    PartialTensor doesn't provide any Tensor like operations but is a
+    wrapper providing the Tensor representing the local partial shard.
+
+    We assume the size of each local tensor to be exactly the same.
+
+    Users can apply custom distributed sharded computations on top of
+    this primitive.
+
+    Args:
+        local_partial_shard (Tensor): Partial result stored across ranks.
+        pg (ProcessGroup): The process group to aggregate on.
+        aggr_op (distributed_c10d.ReduceOp): Way to aggregate the partial result.
+            Default: ``distributed_c10d.ReduceOp.SUM``
+
+    Examples:
+        >>> # All tensors below are of torch.int64 type.
+        >>> # We have 2 process groups, 2 ranks.
+        >>> tensor = torch.arange(2, dtype=torch.int64) + 1 + 2 * rank
+        >>> tensor = torch.stack([tensor, tensor])
+        >>> tensor
+        tensor([[1, 2], [1, 2]]) # Rank 0
+        tensor([[3, 4], [3, 4]]) # Rank 1
+        >>> partial_tensor = _PartialTensor(tensor, distributed_c10d.ReduceOp.MAX)
+        >>> sharding_dim = 0
+        >>> collect_spec = ChunkShardingSpec(
+                dim=sharding_dim,
+                placements=[
+                    "rank:0/cuda:0",
+                    "rank:1/cuda:1",
+                ],
+            )
+        >>> complete_tensor = partial_tensor.reshard(collect_spec)
+        >>> complete_tensor
+        tensor([3, 4]) # Rank 0
+        tensor([3, 4]) # Rank 1
```

Comment:
If the inputs are

```python
tensor([[1, 2], [3, 4]]) # Rank 0
tensor([[5, 6], [7, 8]]) # Rank 1
```

would I be correct that the output will then be:

```python
tensor([5, 6]) # Rank 0
tensor([7, 8]) # Rank 1
```

If so, shall we update the example. It might help make it more explicit.

Reply:
curious, in real-world applications, when are we going to use min/max/prod ops? (I think I can understand the SUM use case :) )

Reply:
If inputs are:
`tensor([[1, 2], [3, 4]]) # Rank 0
tensor([[5, 6], [7, 8]]) # Rank 1`

The outputs will be:
`tensor([6, 8]) # Rank 0
tensor([10, 12]) # Rank 1`.

I will update the comment to make it clearer.

Reply:
One thing I can think of is the max/min pooling might leverage this?

Reply:
> The outputs will be:
> tensor([6, 8]) # Rank 0 tensor([10, 12]) # Rank 1.

I might have missed sth. I saw the example was using `ReduceOp.MAX` instead of `ReduceOp.SUM`. In that case, will it still do a sum?

Reply:
Oh, yes. In the case of `MAX`, your understanding is correct. I will update the comment.